infra/Pos/Network/CLI.hs

```diff
@@ -0,0 +1,196 @@
+      return $ T.TopologyStatic selfType (peersFromList selfPeers)
+  where
+    mkPeers :: DNS.Resolver -> Y.NodeRoutes -> IO [(T.NodeType, Alts NodeId)]
+    mkPeers resolver (Y.NodeRoutes routes) = mapM (mkAlts resolver) routes
+
+    mkAlts :: DNS.Resolver -> Alts NodeName -> IO (T.NodeType, Alts NodeId)
+    mkAlts _ [] = throwM $ EmptyListOfAltsFor self
+    mkAlts resolver alts@(firstAlt:_) = do
+        -- We silently assume all alternatives have the same type
+        altsType <- Y.nmType <$> metadataFor allPeers firstAlt
+        (altsType,) <$> mapM (resolveNodeName cfg resolver) alts
+
+-- | Resolve node name to IP address
+--
+-- We do this when reading the topology file so that we detect DNS problems
+-- early, and so that we are using canonical addresses (IP addresses) for
+-- node IDs.
+resolveNodeName :: NetworkConfigOpts -> DNS.Resolver -> NodeName -> IO NodeId
+resolveNodeName cfg resolver name = do
+    mAddrs <- DNS.lookupA resolver (nameToDomain name)
+    case mAddrs of
+      Left err            -> throwM $ NetworkConfigDnsError err
+      Right []            -> throwM $ CannotResolve name
+      Right addrs@(_:_:_) -> throwM $ NoUniqueResolution name addrs
+      Right [addr]        -> return $ ipv4ToNodeId cfg addr
+  where
+    nameToDomain :: NodeName -> DNS.Domain
+    nameToDomain (NodeName n) = BS.C8.pack (Text.unpack n)
+
+ipv4ToNodeId :: NetworkConfigOpts -> IPv4 -> NodeId
+ipv4ToNodeId NetworkConfigOpts{..} addr =
+    addressToNodeId (BS.C8.pack (show addr), networkConfigOptsPort)
+
+metadataFor :: Y.AllStaticallyKnownPeers -> NodeName -> IO Y.NodeMetadata
+metadataFor (Y.AllStaticallyKnownPeers allStaticallyKnownPeers) node =
+    case M.lookup node allStaticallyKnownPeers of
+      Nothing       -> throwM $ UndefinedNodeName node
+      Just metadata -> return metadata
+
+parseTopology :: FilePath -> IO Y.Topology
+parseTopology fp = do
+    mTopology <- Yaml.decodeFileEither fp
+    case mTopology of
+      Left  err      -> throwM $ CannotParseNetworkConfig err
+      Right topology -> return topology
+
+{-------------------------------------------------------------------------------
+  Errors
+-------------------------------------------------------------------------------}
+
+-- | Something is wrong with the network configuration
+data NetworkConfigException =
+    -- | We cannot parse the .yaml file
+    CannotParseNetworkConfig Yaml.ParseException
+
+    -- | We use a set of statically known peers but we weren't given the
+    -- name of the current node
+  | NetworkConfigSelfUnknown
+
+    -- | The .yaml file contains a node name which is undefined
+  | UndefinedNodeName NodeName
+
+    -- | The static routes for a node contains an empty list of alternatives
+  | EmptyListOfAltsFor NodeName
+
+    -- | Something went wrong during node name resolution
+  | NetworkConfigDnsError DNS.DNSError
+
+    -- | Could not resolve a node name to an IP address
+  | CannotResolve NodeName
```

Comment:
Actually that applies to _all_ of these exceptions. Added a comment to the type:

```
-- NOTE: Behind NAT nodes are not given an explicit network configuration file,
-- but instead rely on the default topology. These exceptions should never be
-- thrown for behind NAT nodes, as we don't want to prevent the user application
-- from starting up.
```